{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "rKKYapagAmC_"
      },
      "outputs": [],
      "source": [
        "import numpy as np\n",
        "import pandas as pd\n",
        "!wget https://gdc-hub.s3.us-east-1.amazonaws.com/download/GDC-PANCAN.survival.tsv\n",
        "!wget https://tcga-pancan-atlas-hub.s3.us-east-1.amazonaws.com/download/Survival_SupplementalTable_S1_20171025_xena_sp\n",
        "!wget https://gdc-hub.s3.us-east-1.amazonaws.com/download/GDC-PANCAN.basic_phenotype.tsv.gz\n",
        "!wget https://gdc-hub.s3.us-east-1.amazonaws.com/download/GDC-PANCAN.methylation27.tsv.gz\n",
        "!wget https://gdc-hub.s3.us-east-1.amazonaws.com/download/GDC-PANCAN.htseq_fpkm-uq.tsv.gz"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "surv = pd.read_csv('Survival_SupplementalTable_S1_20171025_xena_sp', sep='\\t')\n",
        "meta_df = pd.read_csv('GDC-PANCAN.basic_phenotype.tsv.gz', sep='\\t')\n",
        "meth_cols = pd.read_csv('GDC-PANCAN.methylation27.tsv.gz', compression='gzip', sep='\\t',nrows=0).columns\n",
        "ge_cols = pd.read_csv('GDC-PANCAN.htseq_fpkm-uq.tsv.gz',compression='gzip', sep='\\t', index_col=0, nrows=0).columns\n",
        "surv_gdc = pd.read_csv('GDC-PANCAN.survival.tsv' , sep='\\t')\n",
        "surv_gdc = surv_gdc.set_index('sample')\n",
        "print(surv_gdc.shape)\n",
        "print(meta_df.shape)\n",
        "meta_df = meta_df.set_index('sample')\n",
        "## Metafile analysis\n",
        "## Meta file analysis and label dictionary creation\n",
        "c_list=[]\n",
        "rs_list=[]\n",
        "key_control = ['Blood Derived Normal','Bone Marrow Normal','Buccal Cell Normal','FFPE Scrolls','Solid Tissue Normal' ]\n",
        "key_case = ['Additional - New Primary', 'Additional Metastatic','Human Tumor Original Cells','Metastatic','Primary Blood Derived Cancer - Bone Marrow','Primary Blood Derived Cancer - Peripheral Blood','Primary Tumor','Recurrent Blood Derived Cancer - Bone Marrow','Recurrent Blood Derived Cancer - Peripheral Blood','Recurrent Tumor']\n",
        "meta_df_control = meta_df[meta_df['sample_type'].isin(key_control)]\n",
        "meta_df_case = meta_df[meta_df['sample_type'].isin(key_case)]\n",
        "label_dict = {}\n",
        "s1 = list(set(meta_df_control.index) & set(ge_cols) & set(meth_cols))\n",
        "pan_key_normal = []\n",
        "pan_key_cancer = []\n",
        "for k in s1:\n",
        "  label_dict[k] = \"C\"\n",
        "s2 = list(set(meta_df_case.index) &set(ge_cols) & set(meth_cols))\n",
        "for k in s2:\n",
        "  label_dict[k] = \"RS\"\n",
        "\n",
        "## Finding normal and tumor samples in different cancers\n",
        "##  dis_list --> s_tcga : [{key1:value1,key2:value2}]\n",
        "s_tcga = {}\n",
        "dis_list = meta_df['project_id'].unique()\n",
        "for dis in dis_list:\n",
        "  key_normal = []\n",
        "  key_nan = []\n",
        "  key_cancer = []\n",
        "  meta_df1 = meta_df[meta_df['project_id'] == dis]\n",
        "  meta_df1.reset_index(inplace = True)\n",
        "  s_tcga_i ={}\n",
        "  for i in range(0,meta_df1.shape[0]):\n",
        "    if meta_df1['sample_type'][i] in key_control:\n",
        "      key_normal.append(meta_df1['sample'][i])\n",
        "      pan_key_normal.append(meta_df1['sample'][i])\n",
        "    elif pd.isnull(meta_df1['sample_type'][i]):\n",
        "      key_nan.append(meta_df1['sample'][i])\n",
        "    else:\n",
        "      key_cancer.append(meta_df1['sample'][i])\n",
        "      pan_key_cancer.append(meta_df1['sample'][i])\n",
        "\n",
        "  samples_to_use_i = []\n",
        "  samples_to_use_i = key_normal + key_cancer\n",
        "  key_normal = set(ge_cols) & set(meth_cols) & set(key_normal)\n",
        "  key_normal = list(key_normal)\n",
        "  key_cancer =  set(ge_cols) & set(meth_cols) & set(key_cancer)\n",
        "  key_cancer = list(key_cancer)\n",
        "\n",
        "  pan_key_normal = set(ge_cols) & set(meth_cols) & set(pan_key_normal)\n",
        "  pan_key_normal = list(pan_key_normal)\n",
        "  pan_key_cancer =  set(ge_cols) & set(meth_cols) & set(pan_key_cancer)\n",
        "  pan_key_cancer = list(pan_key_cancer)\n",
        "  s_tcga_i['normal'] = key_normal\n",
        "  s_tcga_i['cancer'] = key_cancer\n",
        "  if len(samples_to_use_i) != 0 :\n",
        "    s_tcga[dis] = s_tcga_i\n",
        "\n",
        "  C_count = len(key_normal)\n",
        "  c_list.append(C_count)\n",
        "  print(f\"Control Samples :{C_count}\")\n",
        "  RS_count = len(key_cancer)\n",
        "  rs_list.append(RS_count)\n",
        "  print(f\"Tumour Samples :{RS_count}\")\n",
        "C_count = len(pan_key_normal)\n",
        "print(f\"Pan Control Samples :{C_count}\")\n",
        "RS_count = len(pan_key_cancer)\n",
        "print(f\"Pan Tumour Samples :{RS_count}\")\n",
        "c_dict = {\n",
        "  \"disease\": dis_list,\n",
        "  \"cancer_samples\": rs_list,\n",
        "  \"normal samples\": c_list\n",
        "}\n",
        "c_df = pd.DataFrame(c_dict)\n",
        "print(c_df)\n"
      ],
      "metadata": {
        "id": "KoMTDl30A8gD"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "## Sample selection\n",
        "samples_to_use = pan_key_normal + pan_key_cancer\n",
        "samples_to_use = list(samples_to_use)\n",
        "print(\"Sample Size\")\n",
        "print(len(samples_to_use))"
      ],
      "metadata": {
        "id": "S8i0W1tQBwNk"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!pip install pyDESeq2\n",
        "from pydeseq2.dds import DeseqDataSet\n",
        "from pydeseq2.ds import DeseqStats\n",
        "from sklearn.feature_selection import VarianceThreshold\n",
        "from sklearn.feature_selection import VarianceThreshold"
      ],
      "metadata": {
        "id": "UWazQwGnBySq"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "## DNA Methylation Data\n",
        "## Reading and Preprocessing....\n",
        "samples_to_use.append(('Composite Element REF'))\n",
        "meth = pd.read_csv('GDC-PANCAN.methylation27.tsv.gz', compression='gzip', sep='\\t' ,usecols =samples_to_use)\n",
        "print('Meth Sample size:')\n",
        "print(len(meth.columns))\n",
        "print(meth.shape)\n",
        "meth = meth.dropna(axis=0)\n",
        "print(meth.shape)\n",
        "meth.set_index(\"Composite Element REF\", inplace = True)\n",
        "meth_df = meth.T\n",
        "print(meth_df.shape)\n",
        "var_thr = VarianceThreshold(threshold = 0.001) #Removing features that are 95% or more similar\n",
        "var_thr.fit(meth_df)\n",
        "concol = [column for column in meth_df.columns\n",
        "          if column not in meth_df.columns[var_thr.get_support()]]\n",
        "meth_df = meth_df.drop(concol,axis=1)\n",
        "print(\"Meth shape after deleting less variance features\")\n",
        "print(meth_df.shape)\n",
        "##Differential Analysis\n",
        "Lab = [label_dict[i] for i in meth_df.index]\n",
        "meth_dff= (meth_df * 1000).astype(int)\n",
        "metadata = pd.DataFrame(zip(meth_dff.index,Lab),columns=('Sample','Condition'))\n",
        "print(metadata.groupby('Condition')['Sample'].nunique())\n",
        "metadata=metadata.set_index('Sample')\n",
        "dds = DeseqDataSet(counts = meth_dff, metadata = metadata, design_factors = 'Condition')\n",
        "print(dds)\n",
        "dds.deseq2()\n",
        "stat_res = DeseqStats(dds, contrast = ('Condition','RS','C'))\n",
        "print(stat_res.summary())\n",
        "res = stat_res.results_df\n",
        "m = res[(res.padj < 0.05) & (abs(res.log2FoldChange)>0.5)]\n",
        "print(\"Significant Genes\")\n",
        "print(m)\n",
        "hyper = m[m['log2FoldChange']>0]\n",
        "hypo = m[m['log2FoldChange']<0]\n",
        "hypo_cpg = list(hypo.index)\n",
        "hyper_cpg = list(hyper.index)\n",
        "print(\"Hyper-length\")\n",
        "print(len(hyper_cpg))\n",
        "print(\"Hypo-length\")\n",
        "print(len(hypo_cpg))\n"
      ],
      "metadata": {
        "id": "xPfgqpkbB1cx"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "## Gene Expression Data\n",
        "## Reading and Preprocessing....\n",
        "samples_to_use.append('xena_sample')\n",
        "ge = pd.read_csv('GDC-PANCAN.htseq_fpkm-uq.tsv.gz', compression='gzip', index_col=0, sep ='\\t',usecols =samples_to_use)\n",
        "ge = ge.T\n",
        "print(ge.shape)\n",
        "var_thr = VarianceThreshold(threshold = 0.001)\n",
        "var_thr.fit(ge)\n",
        "concol = [column for column in ge.columns\n",
        "          if column not in ge.columns[var_thr.get_support()]]\n",
        "ge_df = ge.drop(concol,axis=1)\n",
        "print(\"Meth shape after deleting less variance features\")\n",
        "print(ge_df.shape)\n",
        "drop_col=[]\n",
        "for i in ge_df.columns:\n",
        "  count = (ge_df[i] == 0).sum()\n",
        "  if count>(ge_df.shape[0]*0.50):\n",
        "    drop_col.append(i)\n",
        "print(len(drop_col))\n",
        "ge_df=ge_df.drop(columns=drop_col)\n",
        "print(\"GE shape after deleting zero columns:\")\n",
        "print(ge_df.shape)\n",
        "ge_df = ge_df.dropna(axis=1)\n",
        "print(\"GE shape after deleting na\")\n",
        "print(ge_df.shape)\n",
        "## Differential Analysis\n",
        "Labels = [label_dict[y] for y in ge_df.index]\n",
        "metadataG = pd.DataFrame(zip(ge_df.index,Labels),columns=('Sample','Condition'))\n",
        "print(metadataG.groupby('Condition')['Sample'].nunique())\n",
        "metadataG=metadataG.set_index('Sample')\n",
        "ge1 =(ge_df).astype(int)\n",
        "dds = DeseqDataSet(counts = ge1, metadata = metadataG, design_factors = 'Condition')\n",
        "print(dds)\n",
        "dds.deseq2()\n",
        "stat_res = DeseqStats(dds, contrast = ('Condition','RS','C'))\n",
        "print(stat_res.summary())\n",
        "res = stat_res.results_df\n",
        "g = res[(res.padj < 0.05) & (abs(res.log2FoldChange)>0.5)]\n",
        "print(\"Significant Genes\")\n",
        "print(g)\n",
        "up = g[g['log2FoldChange']>0]\n",
        "down = g[g['log2FoldChange']<0]\n",
        "up_id = list(up.index)\n",
        "down_id = list(down.index)\n",
        "print(\"Up-length\")\n",
        "print(len(up_id))\n",
        "print(\"Down-length\")\n",
        "print(len(down_id))\n"
      ],
      "metadata": {
        "id": "xpSoj7WrCRU4"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "### Creating dictionaries\n",
        "!wget https://gdc-hub.s3.us-east-1.amazonaws.com/download/gencode.v22.annotation.gene.probeMap\n",
        "!wget https://gdc-hub.s3.us-east-1.amazonaws.com/download/illuminaMethyl27_hg38_GDC\n",
        "map={}\n",
        "meth_map = pd.read_csv('illuminaMethyl27_hg38_GDC', sep='\\t')\n",
        "map_dict_m = dict(zip(meth_map['#id'], meth_map['gene']))\n",
        "map_mm = pd.read_csv('cpg_methylation_cpg_to_annotation.tsv', sep='\\t')\n",
        "map_dict_mm = dict(zip(map_mm['CpG_id'], map_mm['Symbol']))\n",
        "for (i,j) in map_dict_mm.items():\n",
        "  if pd.isnull(j) or j=='.':\n",
        "    map[i]=map_dict_m[i]\n",
        "  else:\n",
        "    k = map_dict_m[i]\n",
        "    k_new = list(k.split(','))\n",
        "    h = map_dict_mm[i]\n",
        "    h_new = list(h.split(','))\n",
        "    unique_values = set(k_new + h_new)\n",
        "    if '.' in unique_values:\n",
        "      unique_values.remove('.')\n",
        "    map[i]=unique_values\n"
      ],
      "metadata": {
        "id": "-vV7MecNCq_k"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "g_s_me=[]\n",
        "to_drop=[]\n",
        "cpg_to_gs={}\n",
        "gs_to_cpg={}\n",
        "gs_to_cpgg={}\n",
        "mindex = list(m.index)\n",
        "gindex = list(g.index)\n",
        "for i in m.index:\n",
        "  update = map[i]\n",
        "  cpg_to_gs[i] = update\n",
        "  for e in update:\n",
        "    g_s_me.append(e)\n",
        "    if e in gs_to_cpg.keys():\n",
        "      gs_to_cpg[e]= gs_to_cpg[e] +','+ i\n",
        "    else:\n",
        "      gs_to_cpg[e] = i\n",
        "gs_to_gid={}\n",
        "gid_to_gs={}\n",
        "g_s_ge=[]\n",
        "ge_map = pd.read_csv('gencode.v22.annotation.gene.probeMap', sep='\\t')\n",
        "map_dict_r = dict(zip(ge_map['gene'], ge_map['id']))\n",
        "map_dict = dict(zip(ge_map['id'], ge_map['gene']))\n",
        "for i in gindex:\n",
        "  gid_to_gs[i] = map_dict[i]\n",
        "  gs_to_gid[map_dict[i]]=i\n",
        "  g_s_ge.append(map_dict[i])\n",
        "## Overlapping...\n",
        "overlap_g_s = set(g_s_me)&set(g_s_ge)\n",
        "overlap_g_s=list(overlap_g_s)\n",
        "print(\"Length of overlap genes:\")\n",
        "print(len(overlap_g_s))\n",
        "print(\"Overlapped gene symbols:\")\n",
        "print(overlap_g_s)"
      ],
      "metadata": {
        "id": "ZnqEL_TsC6jL"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}